readme.md
### Prácticas PPSD - [[Changelog]](https://github.com/carloshermida/ppsd-practicas/blob/main/changelog.md#changelog)
##### Curso 2021/2022

Este repositorio contiene las prácticas de la asignatura _Protección, Privacidad y Seguridad de Datos_ del grado en _Ciencia e Ingeniería de Datos_ de la _Universidad de Coruña_.

---

#### Monoalfabeto
Creación e implementación de un algoritmo de cifrado simétrico por sustitución monoalfabeto
* Cipher
```
En primer lugar, generamos las listas de consonantes y vocales originales,
y a partir de la semilla obtenemos los n-gramas por los que vamos a sustituir
las letras del texto en claro. Intercambiamos las vocales por su correspondiente
trigrama y las consonantes por su respectivo digrama. Finalmente, añadimos un
espacio cada 4 posiciones para tratar de dificultar el análisis.

Ejemplo:
Texto en claro: “Este es el texto en claro”. 
Semilla: 4

El algoritmo automáticamente pone el texto como “ESTEESELTEXTOENCLARO”
(con la función auxiliar text_format), y con la semilla que le pasamos crea las
siguientes asignaciones para las letras:

Para las vocales: ['AQI', 'IPE', 'OFA', 'EBU', 'UJO']
Para las consonantes: ['QJ', 'PK', 'FL', 'BC', 'JH', 'GB', 'HQ', 'CP', 'MM', 'NF', 'TT',
'SR', 'YV', 'LD', 'KN', 'RW', 'VX', 'DS', 'XZ', 'ZG', 'WY']
 
Finalmente, sustituye en el texto en claro, pone los espacios y devuelve:
IPEK NRWI PEIP EKNI PEMM RWIP EXZR WEBU IPET TPKM MAQI LDEB U
```
* Decipher
```
Esta función comienza ignorando los espacios y, a continuación, si el elemento que
se está seleccionando es una consonante avanza dos elementos y añade un espacio
(debido a que las consonantes son digramas), mientras que si el elemento seleccionado
es una vocal se avanza 3 elementos y se añade un espacio (debido a que las vocales
son trigramas).

Una vez tengamos estos elementos separados, los cambiamos por su correspondiente
letra en claro.

Ejemplo:
Texto cifrado: “IPEK NRWI PEIP EKNI PEMM RWIP EXZR WEBU IPET TPKM MAQI LDEB U”.
Semilla: 4

El algoritmo automáticamente podría el texto como
“IPEKNRWIPEIPEKNIPEMMRWIPEXZRWEBUIPETTPKMMAQILDEBU”, y con la semilla que le pasamos,
lo que hace es crear las siguientes asignaciones para las letras:

Para las vocales: ['AQI', 'IPE', 'OFA', 'EBU', 'UJO']
Para las consonantes: ['QJ', 'PK', 'FL', 'BC', 'JH', 'GB', 'HQ', 'CP', 'MM', 'NF',
'TT', 'SR', 'YV', 'LD','KN', 'RW', 'VX', 'DS', 'XZ', 'ZG', 'WY']

Finalmente, sustituye en el texto cifrado, y devuelve: “ESTEESELTEXTOENCLARO”
```
* Random Shuffle
```
Esta función pretende sustituir a la función random.shuffle de la librería Random.

El objetivo de nuestra función es mezclar de manera aleatoria mediante una semilla
los elementos que le pasamos en forma de array, de manera que con una semilla
siempre se va a hacer de la misma forma.

Por lo tanto, a esta función se le pasan los argumentos seed, que sería un número
entero,y array, que sería la lista de elementos que queremos mezclar.

En primer lugar, se crea una lista vacía array_indexes que guardará los números
del 0 al 9 tantas veces como sean necesarias para que la longitud de array_indexes
sea mayor o igual a la longitud de “array”, que es lo que se hace en el primer bucle.
Además, se inicializa el número random_num a partir de la semilla y una serie
de cálculos.

En el bucle while se va eliminando cada cifra de random_num de array_indexes
(en el caso en el que no queden de esas cifras en el array, simplemente
se pasa a la siguiente).

Si random_num consiguió vaciar array_indexes, ese número es válido. En caso
contrario, se le aplican unos cálculos a este número para obtener otro más largo.

Se crea una lista vacía que va a guardar los índices del array pero mezclados,
shuffled_array_indexes, y la manera de mezclarlos se define en el siguiente bucle,
que lo que hace es coger cada cifra del random_num y añadir ese índice a
shuffled_array_indexes en caso de que ya esté añadido, le suma 10. En caso de que
ese número sea mayor a la longitud del “array” (el que se quiere mezclar),
ese índice se descarta.

Por último, coloca los valores de “array” en las posiciones que indica
shuffled_array_indexes y obtenemos de esta manera el array barajado.

Esta función no consigue para todas las semillas una mezcla diferente, pero lo
hace considerablemente bien. De hecho, como nosotros al crear los abecedarios
permutados utilizamos una semilla y esa misma multiplicada por 3, 5 y 7, es 
muy complicado que casualmente coincidan los abecedarios barajados resultantes.

Ejemplo:
array = [‘a’, ‘b’, ‘c’, ‘d’, ‘e’, ‘f’, ‘g’, ‘h’, ‘i’, ‘j’, ‘k’]
seed = 43 
array_indexes=[0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9]

Primer random_num que crea:
random_num = 16684 
array_indexes=[0,2,3,5,7,9,0,1,2,3,4,5,7,8,9]

A continuación, haría cálculos para obtener otro “random_num”

Último random_num que crea:
random_num = 200096949731595398487625349157381203732 
array_indexes = [ ]
shuffled_array_indexes = [2, 0, 10, 9, 6, 4, 7, 3, 1, 5, 8]
array = [‘c’, ‘a’, ‘k’, ‘j’, ‘g’, ‘e’, ‘h’, ‘d’, ‘b’, ‘f’, ‘i’]

Ese sería el array barajado
```
* Abcd Generator
```
Esta función devolverá:
   clear_cons: una lista con todas las consonantes del abecedario.
   clear_vowel: una lista con todas las vocales del abecedario.
```	
* Key Generator
```
Esta función crea cuatro listas: 2 con las consonantes permutadas y 2 con las
vocales permutadas (con semillas diferentes). Posteriormente, crea trigramas para
referirse a las vocales y digramas para referirse a las consonantes.

Los trigramas de las vocales tienen la estructura de vocal-consonante-vocal 
mientras que los digramas de consonantes tiene la estructura de consonante-consonante.

Esta función devolverá:
	vowel_key: lista con los trigramas que identifican a las vocales. 
	cons_key: lista con los digramas que identifican a los consonantes.

```
* Text format
```
Elimina o cambia todos los caracteres no permitidos en la práctica del texto en
claro que se desea cifrar, como letras con tildes, signos de puntación, espacios etc.
```
* Front-end
```
Para facilitar el uso del algoritmo, creamos una interfaz para el usuario. 
Simplemente se debe introducir el número del apartado al que se quiere acceder
y seguir las instrucciones.
	
	(1) Cifrar
		(1) Escribir el texto
		(2) Importar desde un archivo 
		(0) Volver al menú principal

	(2) Descifrar
		(1) Escribir el texto
		(2) Importar desde un archivo 
		(0) Volver al menú principal

	(0) Salir
```
###### Sintaxis de uso
```
Ejecuta el fichero de Python en cualquier terminal ($python3 monoalfabeto.py).
En el menú principal, si quieres cifrar pulsa el número 1, y si quieres descifrar
el número 2. A continuación, la tecla enter.

Para importar el texto desde un fichero, selecciona la opción 2 e introduce el
nombre del archivo (en caso de que se encuentre en un directorio distinto al
del fichero de Python, se debe introducir la ruta).

Si prefieres escribir el texto, selecciona la opción 1, escribe el texto,
y al terminar confirma que el texto es correcto.

Introduce la clave, teniendo en cuenta que en este caso es numérica (la semilla).

El resultado se muestra por pantalla, y se da la opción de guardarlo en un archivo.
En caso afirmativo, se debe indicar el nombre y extensión. Finalmente, pulsa enter
para volver al menú principal.
```

---

#### Test Random
Programa para comparar el rendimiento de la función *random.shuffle* de la librería de Python _"random"_ y la función *random_shuffle* creada para el algoritmo de cifrado simétrico por sustitución monoalfabeto

---

#### Vigenère
Implementación del algoritmo de Vigenère
* Cipher
```
Esta función crea los abecedarios con desplazamiento en función de la clave introducida. 
A continuación, decide que abecedario le corresponde a cada letra según su posición en
el texto y sustituye por la letra que ocupa el lugar correspondiente a la letra original.

Ejemplo:
Texto en claro: “Este es el texto en claro”
Clave: “cable”

El algoritmo automáticamente pone el texto como “ESTEESELTEXTOENCLARO” (con la función
auxiliar text_format).

abc: [A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z]

abc_list [0]: [C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, A, B] 
abc_list [1]: [A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z] 
abc_list [2]: [B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, A] 
abc_list [3]: [L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, A, B, C, D, E, F, G, H, I, J, K] 
abc_list [4]: [E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, A, B, C, D]

Texto cifrado: “GSUPIUEMEIZTPPRELBCS”
```
* Decipher
```
Esta función crea los abecedarios con desplazamiento en función de la clave introducida. 
A continuación, decide que abecedario le corresponde a cada letra según su posición en el
texto y sustituye por la letra real que ocupa el lugar correspondiente a la letra cifrada
en cada abecedario.

Ejemplo:
Texto cifrado: “GSUPIUEMEIZTPPRELBCS”
Clave: “cable”

abc: [A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z]

abc_list [0]: [C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, A, B] 
abc_list [1]: [A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z]
abc_list [2]: [B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, A] 
abc_list [3]: [L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, A, B, C, D, E, F, G, H, I, J, K] 
abc_list [4]: [E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, A, B, C, D]

Texto en claro: “ESTEESELTEXTOENCLARO”
```
* Abcd Generator
```
Tras obtener el abecedario normal, crea diferentes abecedarios a partir de la clave
introducida (tantos como caracteres tenga la clave). Cada uno de estos abecedarios
empiezan por las letras de la clave y van hasta la letra Z, y vuelven a la A hasta
llegar a la letra anterior a la letra de la clave.
```	
* Text format
```
Elimina o cambia todos los caracteres no permitidos en la práctica del texto en claro
que se desea cifrar, como letras con tildes, signos de puntación, espacios etc.
```
* Front-end
```
Para facilitar el uso del algoritmo, creamos una interfaz para el usuario. 
Simplemente se debe introducir el número del apartado al que se quiere acceder
y seguir las instrucciones.
	
	(1) Cifrar
		(1) Escribir el texto
		(2) Importar desde un archivo 
		(0) Volver al menú principal

	(2) Descifrar
		(1) Escribir el texto
		(2) Importar desde un archivo 
		(0) Volver al menú principal

	(0) Salir
```
###### Sintaxis de uso
```
Ejecuta el fichero de Python en cualquier terminal ($python3 vigenere.py).
En el menú principal, si quieres cifrar pulsa el número 1, y si quieres descifrar
el número 2. A continuación, la tecla enter.

Para importar el texto desde un fichero, selecciona la opción 2 e introduce el
nombre del archivo (en caso de que se encuentre en un directorio distinto al del
fichero de Python, se debe introducir la ruta).

Si prefieres escribir el texto, selecciona la opción 1, escribe el texto, y al
terminar confirma que el texto es correcto.

Introduce la clave, teniendo en cuenta que en este caso está compuesta por letras.

El resultado se muestra por pantalla, y se da la opción de guardarlo en un archivo.
En caso afirmativo, se debe indicar el nombre y extensión. Finalmente, pulsa enter
para volver al menú principal.
```

---

#### RC4
Implementación del algoritmo RC4
* KSA
```
Inicializa la lista S con los números del 0:255, en binario. Genera la lista T,
donde se repite la clave en bucle hasta llegar a 256 elementos.

Finalmente, permuta el vector S.
```
* PRGA
```
Por cada carácter que se quiera cifrar, se realiza una llamada a la función PRGA. 
Esta realiza permutaciones en S y obtiene los valores del keystream. Estos números,
que se guardan en binario, se utilizarán posteriormente para la operación XOR con el
carácter que se quiera cifrar/descifrar.
```
* Cipher
```
En primer lugar, se genera la lista S y se le aplica la permutación inicial.
Creamos el DataFrame donde se guardarán los datos que se mostrarán al usuario posteriormente.
Inicializamos también el keystream, que empieza vacío, y los valores de i y j (0,0).

Procesamos cada carácter que introduzca el usuario y mostramos la codificación y los
cambios que se producen en S (en rojo los valores de S[i] y S[j] que se intercambian,
y en verde el valor que se añade al keystream y que se utilizará para el XOR).
```	
* Decipher
```
En primer lugar, se genera la lista S. Inicializamos también el keystream,
que empieza vacío, y los valores de i y j (0,0).

Automáticamente, se descifra cada carácter del texto cifrado (introducido en hexadecimal),
calculando los valores del keystream y el resultado del XOR. Una vez que tenemos el
texto descifrado completamente, se muestra por pantalla.
```
* Front-end
```
Para facilitar el uso del algoritmo, creamos una interfaz para el usuario. 
Simplemente se debe introducir el número del apartado al que se quiere acceder
y seguir las instrucciones.
	
	(1) Cifrar

	(2) Descifrar

	(0) Salir
```
###### Sintaxis de uso
```
Ejecuta el fichero de Python en cualquier terminal ($python3 RC4.py).

En el menú principal, si quieres cifrar pulsa el número 1, y si quieres
descifrar el número 2. A continuación, la tecla enter.

En el caso del cifrado, se pedirá que se introduzca la clave en hexadecimal.
Después se mostrará por pantalla el estado inicial de la lista S y su primera permutación.
Para continuar, simplemente se debe introducir un carácter y pulsar enter. Se verá una
tabla con todos los datos relevantes de la codificación y los cambios producidos en la
lista S (en rojo los valores de S[i] y S[j] que se intercambian, y en verde el valor que
se añade al keystream y que se utilizará para el XOR). Para terminar la ejecución del
cifrado, se debe introducir la palabra exit y después pulsar enter.

En el caso del descifrado, escribe el texto en hexadecimal, y al terminar confirma
que el texto es correcto. Después, se pedirá que se introduzca la clave en hexadecimal,
y ya se podrá ver en la pantalla el texto descifrado. Finalmente, pulsa enter para
volver al menú principal.
```

---

#### Recursos
Contiene dos pares de archivos cifrados y descifrados para probar los algoritmos

---

#### EasterEggs
Al cerrar cualquiera de los tres programas aparece un texto que está cifrado
según el algoritmo que corresponda. Para descifrarlos, se debe utilizar el código
usando las siguientes claves:
* Monoalfabeto: 2022
* Vigènere: PPSD
* RC4: 7e6 (hex(2022))


changelog.md
### Changelog

---

###### 13/09/22
####  Readme y Changelog - _last commit_
* Implementación del readme para resumir el contenido del repositorio
* Utilización de este changelog para explicar los cambios realizados

---

###### 13/09/22
#### upload - [82bcbce](https://github.com/carloshermida/ppsd-practicas/commit/82bcbce16d988c1bcef7919a53956087ab5ddf0c)
* Subida de las prácticas
* Implementación y creación de los algoritmos requeridos
* Comprobación del correcto funcionamiento de los algoritmos



test_random.py

# CÁLCULO DE PROBABILIDAD DE REPETECIÓN DEL MEZCLADO ALEATORIO
# 			 RANDOM_SHUFFLE VS RANDOM.SHUFFLE
#


##### ABECEDARIO

def abcd_generator(op):
	abc = []
	vow =["A", "E", "I", "O", "U"]
	for a in range(65, 91):
		if op == "all":
			abc.append(chr(a))
		else:
			if chr(a) not in vow:
				abc.append(chr(a))
	return abc


##### TRANSFORMADOR DE ARRAY EN STRING

def merge_list(array):
	string = ""
	for item in array:
		string += item
	return string


##### MEZCLADOR DE LISTAS DEL USUARIO

def random_shuffle(seed, array):
	array_indexes = []
	for rep in range((len(array)//10)+1):
		for i in range(10):
			array_indexes.append(str(i))

	random_num = seed * 97 * len(str(seed))**2
	while True:
		indexes = array_indexes.copy()
		for figure in str(random_num):
			if figure in indexes:
				indexes.remove(figure)
		if len(indexes) == 0:
			break
		random_num = int(str(random_num)[::-1]) * 83

	shuffled_array_indexes = []
	for num in str(random_num):
		while num in shuffled_array_indexes:
			num = str(int(num)+10)
		if int(num) < len(array):
			shuffled_array_indexes.append(num)

	shuffled_array = [0]*len(array)
	for i in range(len(array)):
		shuffled_array[int(shuffled_array_indexes[i])] = array[i]

	return shuffled_array


##### TEST

vow =["A", "E", "I", "O", "U"]
cons = abcd_generator("cons")
abc = abcd_generator("all")

################# PARÁMETROS ###########################
s_rep = 1
e_rep = 20000
use = cons
########################################################

import math
rep = e_rep-s_rep+1
print("\n CÁLCULO DE PROBABILIDAD DE REPETECIÓN DEL MEZCLADO ALEATORIO\n\t       RANDOM_SHUFFLE VS RANDOM.SHUFFLE")
print(f"\nRango de semillas {s_rep}:{e_rep} ({rep} iteraciones)")
print(f"Posibles combinaciones distintas: {math.factorial(len(use))}\n")

## Mezclador creado

cnt = 0
arrays_array = []
for i in range(s_rep, e_rep+1):
	tmp = merge_list(random_shuffle(i, use))
	if tmp in arrays_array:
		cnt +=1
	arrays_array.append(tmp)

print("USUARIO:   {}  {}%".format(cnt, round(cnt/rep, 3)*100))


## Mezclador de la librería random

import random
cnt = 0
arrays_array = []
for i in range(s_rep, e_rep+1):
	tmp_abc = use.copy()
	random.seed(i)
	random.shuffle(tmp_abc)
	tmp = merge_list(tmp_abc)
	if tmp in arrays_array:
		cnt +=1
	arrays_array.append(tmp)

print("LIBRERÍA:  {}  {}%".format(cnt, round(cnt/rep, 3)*100))


######

# Esta es la probablidad (2.5% - 4%) de que se repita una combinacion cogiendo semillas secuenciales
# Sin embargo, en nuestro algoritmo utilizamos la seed, seed*3, seed*5, seed*7; lo que disminuye la probabilidad
# de que se mezclen de igual manera las listas. Además, si por ejemplo la semilla 6 y la semilla 43 mezclaran de la
# misma manera, las siguientes listas que multiplican la semilla por los numeros descritos anteriormente no serían las mismas.

######








RC4.py
#
#                             ALGORITMO RC4
# 
#     Nina López | Aitana Martínez | Patricia Da Concepción | Carlos Hermida


######## LIBRERÍAS

# Sólamente las usamos para la parte de front-end
# Es totalmente independiente a la implementación de RC4

import pandas as pd
import os
from time import sleep
import sys


######## KSA

def KSA(key, log = True):
    
    # Inicialización
    S = [bin(x)[2:] for x in range(256)]
    if log:
        print("\nInicialización de S: \n\n{}".format(' '.join(S)))
    T = [key[x % len(key)] for x in range(256)]
    
    # Permutación de S
    j = 0
    for i in range(256):
        j = (j + int(S[i],2) + int(T[i], 16)) % 256
        S[i], S[j] = S[j], S[i]
    if log:
        print("\n\nPermutación inicial de S: \n\n{}".format(' '.join(S)))
    return S


######## PRGA

def PRGA(S, keystream, i, j):
    i = (i + 1) % 256
    j = (j + int(S[i],2)) % 256
    S[i], S[j] = S[j], S[i]
    t = (int(S[i],2) + int(S[j],2)) % 256
    keystream.append(S[t])
    return S, keystream, i, j


######## CIFRADO

def cipher(key):
    
    # Introducción de la clave y KSA
    key = key.split(" ")
    S = KSA(key)

    # Inicialización del DataFrame, keystream (vacío) e i,j iniciales (0,0)
    df = pd.DataFrame(columns=["LETRA", "ASCII(10)", "ASCII(2)", "KEYSTREAM", "CIFRADO(2)", "CIFRADO(16)"])
    keystream = []
    i, j = 0, 0

    while True:

        # Introducción del texto letra por letra
        letter = input("\n> ")
        if letter == "exit":
            break
        if len(letter) != 1:
            continue

        # PRGA y XOR
        S, keystream, i, j = PRGA(S, keystream, i, j)
        cipher_letter = int(keystream[-1],2)^ord(letter)
    
        # Acualización del DataFrame
        df.loc[len(df.index)] = [letter, ord(letter), bin(ord(letter))[2:], keystream[-1], bin(cipher_letter)[2:], hex(cipher_letter)[2:]]
        print(df)
    
        # Muestra del estado de S
        print("\n\nModificación de S: \n")
        print(f"S[{i}] <-> S[{j}] | keystream -> S[{S.index(keystream[-1])}]\n")
        s_show = ""
        for element in S:
            if S.index(element) in [i,j]:
                s_show += ("\033[1;31m{}\033[00m").format(element)
            elif element == keystream[-1]:
                s_show += ("\033[1;32m{}\033[00m").format(element)
            else:
                s_show += element
            s_show += " "
        print(s_show)


######## DESCIFRADO

def decipher(cipher_text, key):

    # Introducción del texto cifrado
    cipher_text = cipher_text.split(" ")
    
    # Introducción de la clave y KSA
    key = key.split(" ")
    S = KSA(key, log = False)

    # Inicialización del keystream (vacío) e i,j iniciales (0,0)
    keystream = []
    clear_text = ""
    i, j = 0, 0

    for element in cipher_text:
        # PRGA y XOR
        S, keystream, i, j = PRGA(S, keystream, i, j)
        clear_text += chr(int(keystream[-1],2)^int(element, 16))

    return clear_text


######## FRONT-END

### CLEAR

def clear():
    # for windows
    if os.name == 'nt':
        os.system('cls')
    # for mac and linux
    else:
        os.system('clear')


### MENÚ PRINCIPAL

def main_menu():
    clear()
    print("\n MENÚ PRINCIPAL\n\n 1. Cifrar\n 2. Descifrar\n 0. Salir\n")
    
    while True:
        try:
            main_choose = int(input("> "))
            break
        except:
            print(" ERROR / Opción incorrecta")
            sleep(2)
            main_menu()

    if main_choose == 1:
        job_menu(1)
    elif main_choose == 2:
        job_menu(2)
    elif main_choose == 0:
        clear()
        print("\n -- c8 65 de f2 c2 24 54 d6 78 b8 5a d2 !")
        sys.exit()   
    else:
        print(" ERROR / Opción incorrecta")
        sleep(2)
        main_menu()


### MENÚ DE CIFRADO / DESCIFRADO

def job_menu(op):
    clear()
    if op == 1:
        print("\n ALGORITMO DE CIFRADO")

    else:
        confirm_op = "n"
        while confirm_op not in ["y", "Y"]:
            print("\n ALGORITMO DE DESCIFRADO")
            cipher_text = input("\n Introduce el texto en hexadecimal a continuación.\n Pulsa [enter] para finalizar.\n\n > ")
            confirm_op = input("\n ¿Estás seguro? (y/n) \n Consejo: copia el texto en el portapapeles si quieres consevarlo o modificarlo\n > ")
            clear()

    while True:
        try:
            key = input("\n Introduce la clave: ")
            assert(len(key) != 0)
            for element in key.split(" "):
                int(element, 16)
            break
        except:
            print(" ¡Recuerda el formato de la clave!")

    if op == 1:
        print("\n Escribe 'exit' para terminar la ejecución\n")
        cipher(key)

    else:
        try:
            output = decipher(cipher_text, key)
            print(" Texto descifrado\n\n", output)
        except:
            print("\n ERROR / Texto cifrado no válido")
            sleep(3)
            main_menu()

    input("\n Pulsa [enter] para volver al menú principal.\n")
    main_menu()


### PANTALLA DE INICIO

def start():
    
    clear()
    logo =("\
  ____   ____ _  _   \n\
 |  _ \\ / ___| || |  \n\
 | |_) | |   | || |_ \n\
 |  _ <| |___|__   _|\n\
 |_| \\_\\\\____|  |_|  \n\
                     \n\n")
       
    print(logo, "\n"*2)
    sleep(2)                                                            
    main_menu()


######## EJECUCIÓN DEL SCRIPT

if __name__ == "__main__":
    start()









monoalfabeto.py
#
#      		ALGORITMO DE CIFRADO SIMÉTRICO POR SUSTITUCIÓN MONOALFABETO
# 
#     Nina López | Aitana Martínez | Patricia Da Concepción | Carlos Hermida


######## LIBRERÍAS

# Sólamente las usamos para la parte de front-end
# El algoritmo de cifrado es totalmente independiente

import os
from time import sleep
import sys


######## RANDOM 

def random_shuffle(seed, array):
	array_indexes = []
	for rep in range((len(array)//10)+1):
		for i in range(10):
			array_indexes.append(str(i))

	random_num = seed * 97 * len(str(seed))**2
	while True:
		indexes = array_indexes.copy()
		for figure in str(random_num):
			if figure in indexes:
				indexes.remove(figure)
		if len(indexes) == 0:
			break
		random_num = int(str(random_num)[::-1]) * 83

	shuffled_array_indexes = []
	for num in str(random_num):
		while num in shuffled_array_indexes:
			num = str(int(num)+10)
		if int(num) < len(array):
			shuffled_array_indexes.append(num)

	shuffled_array = [0]*len(array)
	for i in range(len(array)):
		shuffled_array[int(shuffled_array_indexes[i])] = array[i]

	return shuffled_array
	

######## ABECEDARIO

def abcd_generator():
	clear_cons = []
	clear_vowel = ["A", "E", "I", "O", "U"]

	for code in range(65,91):
		letter = chr(code)
		if letter not in clear_vowel:
			clear_cons.append(letter)

	return (clear_cons, clear_vowel)


######## CLAVES

def key_generator(cons, vowel, seed):
	clear_cons, clear_vowel = abcd_generator()
	cons1 = random_shuffle(seed, clear_cons)
	cons2 = random_shuffle(seed*3, clear_cons)
	vowel1 = random_shuffle(seed*5, clear_vowel)
	vowel2 = random_shuffle(seed*7, clear_vowel)
	vowel_key = []
	cons_key = []
	
	for i in range(len(vowel)):
		vowel_key.append(vowel1[i]+cons1[i]+vowel2[i])

	for i in range(len(cons)):
		cons_key.append(cons1[i]+cons2[i])

	return (vowel_key, cons_key)


######## CIFRADO

def cipher(text, seed):
	clear_cons, clear_vowel = abcd_generator()
	vowel_key, cons_key = key_generator(clear_cons, clear_vowel, seed)
	encrypted_text = ""
	encrypted_spaced_text = ""

	for letter in text:
		if letter in clear_vowel:
			encrypted_text += vowel_key[clear_vowel.index(letter)]
		elif letter in clear_cons:
			encrypted_text += cons_key[clear_cons.index(letter)]

	cnt = 0
	for i in encrypted_text:
		encrypted_spaced_text += i
		cnt += 1
		if cnt % 4 == 0:
			encrypted_spaced_text += " "

	return encrypted_spaced_text


######## DESCIFRADO

def decipher(text, seed):
	clear_cons, clear_vowel = abcd_generator()
	vowel_key, cons_key = key_generator(clear_cons, clear_vowel, seed)
	decrypted_spaced_text = ""
	decrypted_text = ""
	text = text.replace(" ", "")
	i = 0
	
	while i < len(text):
		if text[i] in clear_cons:
			decrypted_spaced_text += text[i]+text[i+1]
			i += 2
		elif text[i] in clear_vowel:
			decrypted_spaced_text += text[i]+text[i+1]+text[i+2]
			i += 3
		decrypted_spaced_text += " "

	for code in decrypted_spaced_text.split(" "):
		if code in vowel_key:
			decrypted_text += clear_vowel[vowel_key.index(code)]
		elif code in cons_key:
			decrypted_text += clear_cons[cons_key.index(code)]

	return decrypted_text


######## FORMATEO DE TEXTO

def text_format(text):
    text = text.upper().replace("\n", "").replace("\t", "").replace(" ", "").replace(".", "")
    text = text.replace("Á", "A").replace("É", "E").replace("Í", "I").replace("Ó", "O").replace("Ú", "U")
    text = text.replace("Ñ", "NH").replace("?", "").replace("¿", "").replace("¡", "").replace("!", "").replace("(", "")
    text = text.replace(",", "").replace(";", "").replace(":", "").replace("-", "").replace(")", "")
    for num in range(10):
        text = text.replace(str(num), "")
    return text


######## FRONT-END

### CLEAR

def clear():
	# for windows
	if os.name == 'nt':
		os.system('cls')
	# for mac and linux
	else:
		os.system('clear')


### MENÚ PRINCIPAL

def main_menu():
	clear()
	print("\n MENÚ PRINCIPAL\n\n 1. Cifrar\n 2. Descifrar\n 0. Salir\n")
	
	while True:
		try:
			main_choose = int(input("> "))
			break
		except:
			print(" ERROR / Opción incorrecta")
			sleep(2)
			main_menu()

	if main_choose == 1:
		job_menu(1)
	elif main_choose == 2:
		job_menu(2)
	elif main_choose == 0:
		clear()
		print("\n -- DBOW IQTR XOWI YPSR IKOX WRXI KO !")
		sys.exit()   
	else:
		print(" ERROR / Opción incorrecta")
		sleep(2)
		main_menu()


### MENÚ DE CIFRADO / DESCIFRADO

def job_menu(op):
	clear()
	if op == 1:
		print("\n ALGORITMO DE CIFRADO\n")
	else:
		print("\n ALGORITMO DE DESCIFRADO\n")

	print(" 1. Escribir texto\n 2. Importar desde un archivo\n 0. Volver al menú principal\n")

	while True:
		try:
			job_choose = int(input("> "))
			break
		except:
			print(" ERROR / Opción incorrecta")
			sleep(2)
			job_menu(op)

	if job_choose == 1:
		confirm_op = "n"
		while confirm_op not in ["y", "Y"]:
			clear()
			input_text = input("\n Introduce el texto a continuación.\n Pulsa [enter] para finalizar.\n\n> ")
			confirm_op = input("\n ¿Estás seguro? (y/n) \n Consejo: copia el texto en el portapapeles si quieres consevarlo o modificarlo\n > ")

	elif job_choose == 2:
		cnt = 3
		while True:
			try:
				file = input("\n Introduce el nombre completo del archivo: ")
				with open (file, "r") as f:
					input_text = f.read()
				break
			except:
				cnt -= 1
				if cnt == 0:
					job_menu(op)
				print(" El archivo no ha sido encontrado o está dañado ({} intento(s) restante(s))".format(cnt))

	elif job_choose == 0:
		main_menu()

	else:
		print(" ERROR / Opción incorrecta")
		sleep(2)
		job_menu(op)

	input_text = text_format(input_text)
	
	clear()
	while True:
		try:
			key = int(input("\n Introduce la clave: "))
			assert(len(str(key)) != 0)
			break
		except:
			print(" ¡Recuerda el formato de la clave!")

	if op == 1:
		output = cipher(input_text, key)
		print("\n Texto cifrado\n", output)

	else:
		try:
			output = decipher(input_text, key)
			assert(len(output) != 0)
		except:
			print("\n ERROR / Texto descifrado incorrectamente")
			sleep(3)
			main_menu()
		print("\n Texto descifrado\n", output)

	save = input("\n ¿Desea guardarlo? (y/n)\n > ")
	if save in ["y", "Y"]:
		file_name = input(" Nombre del archivo: ")
		with open (file_name, "w") as f:
			f.write(output)

	input("\n Pulsa [enter] para volver al menú principal.\n")
	main_menu()


### PANTALLA DE INICIO

def start():
	
	clear()
	logo =("\
		                    _  __       _          _\n\
  _ __ ___   ___  _ __   ___   __ _| |/ _| __ _| |__   ___| |_ ___\n\
 | '_ ` _ \\ / _ \\| '_ \\ / _ \\ / _` | | |_ / _` | '_ \\ / _ \\ __/ _ \\\n\
 | | | | | | (_) | | | | (_) | (_| | |  _| (_| | |_) |  __/ || (_) |\n\
 |_| |_| |_|\\___/|_| |_|\\___/ \\__,_|_|_|  \\__,_|_.__/ \\___|\\__\\___/\n\n")
	   
	print(logo, "\n"*2)
	sleep(2)															
	main_menu()


######## EJECUCIÓN DEL SCRIPT

if __name__ == "__main__":
	start()




